contracts/adapters/CharmAdapter.sol

```diff
@@ -0,0 +1,378 @@
+      IOptionToken token = IOptionToken(options);
+      //get market
+      IOptionMarket market = IOptionMarket(token.market());
+
+      // Get strike index, whether is long
+      AdapterStorageTypes.CharmOptionType memory optionType = adapterStorage.addressToOptionType(options);
+
+      //profit of exercising
+      try optionViews.getSellOptionCost(market, optionType.isLongToken, optionType.strikeIndex, amount) returns (uint v) {
+        return v;
+      } catch {
+        return 0;
+      }
+    }
+
+    function canExercise(
+        address options,
+        uint256,
+        uint256 amount
+    ) public view override returns (bool) {
+        //get token
+        IOptionToken token = IOptionToken(options);
+        //get market
+        IOptionMarket market = IOptionMarket(token.market());
+
+        if (block.timestamp < market.expiryTime()) {
+            return false;
+        }
+        if (exerciseProfit(options, 0, amount) > 0) {
+            return true;
+        }
+        return false;
+    }
+
+    /**
+     * @notice Purchases the options contract.
+     * @param optionTerms is the terms of the option contract
+     * @param amount is the purchase amount in Wad units (10**18)
+     * @param maxCost is the max amount of paymentToken to be paid for the option (to avoid sandwich attacks, ...)
+     */
+    function purchase(
+        ProtocolAdapterTypes.OptionTerms calldata optionTerms,
+        uint256 amount,
+        uint256 maxCost
+    ) external payable override returns (uint256) {
+      require(
+          block.timestamp < optionTerms.expiry,
+          "Cannot purchase after expiry"
+      );
+
+      //get token address
+      address tokenAddress = lookupOToken(optionTerms);
+      if(tokenAddress == address(0)){
+        this.populateOTokenMappings();
+      }
+
+      tokenAddress = lookupOToken(optionTerms);
+
+      require(tokenAddress != address(0), "Market needs to exist!");
+
+      // Get strike index, whether is long
+      AdapterStorageTypes.CharmOptionType memory optionType = adapterStorage.addressToOptionType(tokenAddress);
+      //get token
+      IOptionToken token = IOptionToken(tokenAddress);
+      //get market
+      IOptionMarket market = IOptionMarket(token.market());
+      IERC20 baseToken = IERC20(optionTerms.optionType == ProtocolAdapterTypes.OptionType.Put ? optionTerms.strikeAsset : optionTerms.underlying);
+
+      uint256 amountIn;
+
+      if(baseToken.isETH()){
+        amountIn = market.buy{value: address(this).balance}(
+          optionType.isLongToken,
+          optionType.strikeIndex,
+          amount,
+          maxCost
+        );
+      }else{
+        uint256 premium = optionViews.getBuyOptionCost(market, optionType.isLongToken, optionType.strikeIndex, amount);
+
+        uint256 balanceBefore = baseToken.uniBalanceOf(address(this));
+        baseToken.uniTransferFromSenderToThis(premium);
+        uint256 balanceAfter = baseToken.uniBalanceOf(address(this));
+
```

Comment:
if the option bought is a put, as it stands now a person will have to transfer USDC.

practically speaking if a person buys an exclusively charm wbtc strangle with put / call, they will have to first approve usdc, and then wbtc, and then they are ready to buy it. a lot of friction, is this ok?

Reply:
This is not ideal, because the RibbonVol contract expects to receive payment in ETH. Let's do a Uniswap swap from ETH -> USDC to make the purchase happen. This is sort of what we did for [Opyn](https://github.com/ribbon-finance/structured-products/blob/master/contracts/adapters/GammaAdapter.sol#L205-L221)